gap/pairingheap.gi

```diff
@@ -101,19 +92,60 @@ function(heap)
         fi;
     end;
 
-    merge_pairs := function(isLess, heaps)
-        local h, res;
+    # TODO: Find out whether recursive vs non-recursive
+    #       has any measurable benefits
+
+    # Non-recursive merge-pairs
+    merge_pairs_nr := function(isLess, heaps)
+        local l, res, tmp, k, s, i, r;
 
-        if Length(heaps) = 0 then
+        l := Length(heaps);
+
+        if l = 0 then
             return [0,0,0];
+        elif l = 1 then
+            return heaps[1];
         else
-            res := heaps[1];
-
-            for h in heaps{[2..Length(heaps)]} do
-                res := meld(isLess, res, h);
+            res := heaps;
+            k := l;
+            s := 1;
+
+            while k > 1 do
+                r := RemInt(k, 2);
+                k := QuoInt(k, 2);
```

Comment:
This is of course code that would be tremendously faster if written in C, simply because it'd amount to a simply bit-and plus a shift. That's a key part for the speed of the binary heap.

Alas, this would belong into a separate PR ("optimizing pairing heap", instead of this "fix pairing heap" PR, no?